### PR TITLE
Performence improvments

### PR DESCRIPTION
**Pull request by shadmansaleh**
This reduces startup time.

### Changes :
- Only load default config when needed
- Only load cterm colors when needed
- Async load components and theme

Without these changes according to packers [profiler](https://github.com/wbthomason/packer.nvim#profiling) lualine takes around 73ms in my device with these changes around 12ms . Thats **6x** improvement.

norcallies [profiler](https://github.com/norcalli/profiler.nvim) shows similar improvements.

### CODE REVIEW

**Comment by hoob3rt** [comment]
I like these changes. There is one thing I'm not sure of as far as code quality goes. Why do you define `async loader` this way?
```lua
  local async_loader
  async_loader = vim.loop.new_async(vim.schedule_wrap(function()
```

I get that you are sending `async_loader:close()` inside this function but is there no better way to do this? Won't a stop signal be enough? Also where did you find docs for the neovim api, I cannot find it myself... 😞

**Comment by shadmansaleh** [comment]
> I get that you are sending async_loader:close() inside this function but is there no better way to do this?

I don't really know.

> Won't a stop signal be enough? 

What's that ?

> Also where did you find docs for the neovim api, I cannot find it myself... 😞

By api you mean `vim.api` stuff ?

File: lua/lualine/highlight.lua [hunk 1]
```diff
@@ -1,24 +1,23 @@
 -- Copyright (c) 2020-2021 hoob3rt
 -- MIT license, see LICENSE for more details.
 local M = {}
-local utils_colors = require 'lualine.utils.cterm_colors'
+local utils_colors
```
Comment (hoob3rt) [inline lua/lualine/highlight.lua:4]: I would name this `cterm_colors` since that's all it is really

**Review by hoob3rt — COMMENTED** [review]
> I basically followed example in docs

Okay :smile:

> The luv docs are [here](https://github.com/luvit/luv/blob/master/docs.md)
> uv docs [here](http://docs.libuv.org/en/v1.x/)

Okay, that's what I found as well, for some reason I coudn't find `new_async` in this file an hour ago...

File: lua/lualine/highlight.lua (hunk 1 above)
Comment (shadmansaleh) [reply lua/lualine/highlight.lua]: I agree . Renamed

**Comment by hoob3rt** [comment]
lgtm, this is really nice. 2.5ms -> 0.12ms with my config, thank you sir

**Comment by nifoc** [comment]
I think this breaks setting custom colors for (at least) the diagnostics component.

```
...
color_error = lualine_utils.extract_highlight_colors('LspDiagnosticsSignError', 'fg'),
color_warn = lualine_utils.extract_highlight_colors('LspDiagnosticsSignWarning', 'fg'),
color_info = lualine_utils.extract_highlight_colors('LspDiagnosticsSignInformation', 'fg'),
...
```

Work as expected with 64ab49f55ada99db1e613fc1b9b484921c37c21e, but doesn't use the extracted colors with the current master.

**Comment by hoob3rt** [comment]
> I think this breaks setting custom colors for (at least) the diagnostics component.
> 
> ```
> ...
> color_error = lualine_utils.extract_highlight_colors('LspDiagnosticsSignError', 'fg'),
> color_warn = lualine_utils.extract_highlight_colors('LspDiagnosticsSignWarning', 'fg'),
> color_info = lualine_utils.extract_highlight_colors('LspDiagnosticsSignInformation', 'fg'),
> ...
> ```
> 
> (I use the `extract_highlight_colors` function in my configuration)
> 
> Work as expected with [64ab49f](https://github.com/hoob3rt/lualine.nvim/commit/64ab49f55ada99db1e613fc1b9b484921c37c21e), but doesn't use the extracted colors with the current master.

setting custom colors works fine in my config, even with your snippet. If this issue persists for you then please create a new issue

**Comment by shadmansaleh** [comment]
@nifoc what colorscheme are you useing . After aton of digfing I found out the colorcheme was [setting](https://github.com/folke/tokyonight.nvim/blob/4d7251fe0ade96516a6401f58075df006cbc213e/lua/tokyonight/theme.lua#L337) those highlight groups :P . Lualine works normally :)

**Comment by hoob3rt** [comment]
I had to revert this one as it causes a prety major issue. It basically replicates every component's state from an active window to all other windows, 

<details><summary>take a look at this screenshot</summary>

Screenshot 2021-05-06 at 04 50 48

</details>

This may be resolved with #210, although we would have to set this newly introduces parameter to all components

**Comment by shadmansaleh** [comment]
Take a look at #213

**Pull request merged**